### ISSUE TITLE
`repr(C)` is not respected for enums with uninhabited variants

### ISSUE THREAD

**Issue by jswrenn** [rust-lang/rust#68065]
Uninhabited `repr(C)` enums are size zero, _not_ the size of their discriminant. [For instance](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ab603e5a7bb2d0894ea22595bfdf599e):

```rust
enum Never {}

#[repr(C, u64)]
enum E {
    _X(Never),
    _Y(Never),
}

fn main (){
    use core::mem::size_of;

    let never_size = size_of::<Never>();
    assert_eq!(0, never_size);
    
    let disr_size = size_of::<u64>();
    assert_eq!(8, disr_size);
    
    let expected_enum_size = disr_size + never_size;
    let actual_enum_size = size_of::<E>();
    
    assert_eq!(expected_enum_size, actual_enum_size); // FAIL! 8 != 0
}
```

At best, this leads to nonsensical errors. For instance, [this][infinite-size]:
```rust
enum Never {}

#[repr(C, u64)]
enum E {
    _X(Never),
    _Y(E),
}
```
[produces this error message](infinite-size):
```
error[E0072]: recursive type `E` has infinite size
 --> src/lib.rs:4:1
  |
4 | enum E {
  | ^^^^^^ recursive type has infinite size
5 |     _X(Never),
6 |     _Y(E),
  |        - recursive without indirection
  |
  = help: insert indirection (e.g., a `Box`, `Rc`, or `&`) at some point to make `E` representable
```

[infinite-size]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ef95b35c6f9951ee82edfda0ab6e8f6a

At worst, it violates the specification of [RFC2195: _Really Tagged Unions_](https://rust-lang.github.io/rfcs/2195-really-tagged-unions.html), which dictates [that `E` should be layout-compatible to the `union` of its variants](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=86bf6b539516cc82280ddedb15730c45):
```rust
use core::mem::size_of;

#[derive(Copy, Clone)]
enum Never {}
// This:

#[repr(C, u64)]
enum E {
    _X(Never),
    _Y(Never),
}

// ...should have the same layout as `ERepr`:

#[derive(Copy, Clone)]
#[repr(C)]
union ERepr {
    A: EX,
    B: EY,
}

type ETag = u64;

#[derive(Copy, Clone)] struct EX(ETag, Never);
#[derive(Copy, Clone)] struct EY(ETag, Never);

// ...but it doesn't:

fn main() {
    assert_eq!(size_of::<E>(), size_of::<ERepr>()); // FAIL: 0 != 8
}
```

**Comment by fmease** [comment]
Triage: The first and third snippet no longer fail the asserts:

* Re. 1st snippet: `expected_enum_size` = `8` = `actual_enum_size` => RUN PASS
* Re. 3rd snippet: `size_of::<E>()` = `8` = `size_of::<ERepr>()` => RUN PASS

The second snippet still leads to E0072.

---

Disclaimer: I haven't actually really read the contents of this issue, this is a just a quick drive-by triage as part of mass-triaging old issues.

**Comment by zachs18** [comment]
`repr(C)` and `repr(C, Int)` inhibits the all-absent-variant layout optimization since https://github.com/rust-lang/rust/pull/123043

`repr(Int)` will also inhibit the all-absent-variant layout optimization after  https://github.com/rust-lang/rust/pull/146989

**Comment by zachs18** [comment]
> At best, this leads to nonsensical errors. For instance, [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ef95b35c6f9951ee82edfda0ab6e8f6a):
> 
>     enum Never {}
>     
>     #[repr(C, u64)]
>     enum E {
>         _X(Never),
>         _Y(E),
>     }
> 
> [produces this error message](infinite-size):
> 
> ```
> error[E0072]: recursive type `E` has infinite size
>  --> src/lib.rs:4:1
>   |
> 4 | enum E {
>   | ^^^^^^ recursive type has infinite size
> 5 |     _X(Never),
> 6 |     _Y(E),
>   |        - recursive without indirection
>   |
>   = help: insert indirection (e.g., a `Box`, `Rc`, or `&`) at some point to make `E` representable
> ```

I don't think this error message is related to it being an `enum` or having an uninhabited variant; this happens even for `struct`s that contain themselves: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=ea79769ee374d7c8038b731989b87177)

```Rust
#[repr(C)]
struct E(E);
```

```rs
error[E0072]: recursive type `E` has infinite size
 --> src/lib.rs:2:1
  |
2 | struct E(E);
  | ^^^^^^^^ - recursive without indirection
  |
help: insert some indirection (e.g., a `Box`, `Rc`, or `&`) to break the cycle
  |
2 | struct E(Box<E>);
  |          ++++ +

For more information about this error, try `rustc --explain E0072`.
error: could not compile `playground` (lib) due to 1 previous error
```
